src/Microsoft.SqlTools.ServiceLayer/SqlAssessment/GenerateScriptOperation.cs

```diff
@@ -0,0 +1,164 @@
+﻿//
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+//
+
+using System;
+using System.Diagnostics;
+using System.Globalization;
+using System.Text;
+using System.Threading;
+using Microsoft.SqlTools.ServiceLayer.SqlAssessment.Contracts;
+using Microsoft.SqlTools.ServiceLayer.TaskServices;
+using Microsoft.SqlTools.Utility;
+
+namespace Microsoft.SqlTools.ServiceLayer.SqlAssessment
+{
+    /// <summary>
+    /// Generates a script storing SQL Assessment results to a table.
+    /// </summary>
+    internal sealed class GenerateScriptOperation : ITaskOperation
+    {
+        private CancellationTokenSource cancellation = null;
+
+        /// <summary>
+        /// Gets the unique id associated with this instance.
+        /// </summary>
+        public string OperationId { get; set; }
+
+        /// <summary>
+        /// Gets the parameters containing assessment results
+        /// to be stored in a data table.
+        /// </summary>
+        public GenerateScriptParams Parameters { get; }
+
+        /// <summary>
+        /// Gets or sets the error message text
+        /// if an error occurred during task execution
+        /// </summary>
+        public string ErrorMessage { get; set; }
+
+        /// <summary>
+        /// Gets or sets the sql task that's executing the operation
+        /// </summary>
+        public SqlTask SqlTask { get; set; }
+
+        public GenerateScriptOperation(GenerateScriptParams parameters)
+        {
+            Validate.IsNotNull(nameof(parameters), parameters);
+            Parameters = parameters;
+        }
+
+        /// <summary>
+        /// Execute a task
+        /// </summary>
+        /// <param name="mode">Task execution mode (e.g. script or execute)</param>
+        /// <exception cref="InvalidOperationException">
+        /// The method has been called twice in parallel for the same instance.
+        /// </exception>
+        public void Execute(TaskExecutionMode mode)
+        {
+            // This method is not re-enterable as a trade-off for avoiding
+            // overcomplicated multiple CancellationTokenSources management
```

Comment:
Did you see an instance of this happening? I haven't spent a lot of time in this code but a quick look through makes me wonder whether re-entry is even possible since we create a new operation every time a request comes in.

Reply:
No, this method is called once per instance. That's why I haven't implemented a facility for multiple CancellationTokenSources. But there is no guarantee. I used an exception to catch possible bugs in the future. I'd rather ensure the invariant of `cancellation != null` with GenerateScriptOperation constructor, but ITaskOperation was not IDisposable enough to ensure proper disposal of the CancellationTokenSource.  I'd be glad to remove this check if everyone calling ITaskOperation.Execute was informed of the consequences.

Reply:
Other existing operations just create a cancellation token on operation creation. I'd say just do the same here so at least we're consistent - the current way the code is set up I don't see the need to try and make any special cases for trying to handle that.

Reply:
They are not calling the CancellationTokenSource.Dispose method. Looks odd.

Reply:
When was this method called? [SchemaCompareGenerateScriptOperation.Dispose](https://github.com/microsoft/sqltoolsservice/blob/cf6eec47d3a195f906706668ae1ee327494c2936/src/Microsoft.SqlTools.ServiceLayer/SchemaCompare/SchemaCompareGenerateScriptOperation.cs#L95)

Reply:
Made `GenerateScriptOperation` IDisposable.